test/emmy/tape_test.cljc

```diff
@@ -0,0 +1,495 @@
+                      (is (identical? t (g/simplify t))))))))))
+
+(deftest tape-api-tests
+  (testing "tag-of"
+    (checking "tag-of matches tape-tag for cells" 100 [tag gen/nat]
+              (let [cell (t/make tag 1)]
+                (is (= (t/tag-of cell)
+                       (t/tape-tag cell))
+                    "for tape cells, these should match")))
+
+    (checking "for any other type tag == ##-Inf" 100 [x gen/any]
+              (is (= ##-Inf (t/tag-of x))
+                  "for tape cells, these should match")))
+
+  (checking "deep-primal returns nested primal" 100 [p gen/any-equatable]
+            (let [cell (t/make 0 (t/make 1 p))]
+              (is (= p (t/deep-primal cell))
+                  "for tape cells, these should match"))))
+
+(deftest reverse-mode-tests
+  (testing "topological-sort"
+    (let [x (t/make 0 'x)
+          y (t/make 0 'y)
+          a (g/mul x y)
+          b (g/sin x)
+          z (g/add a b)]
+      (is (= [z b a y x]
+             (t/topological-sort z))
+          "topological-sort returns the graph in reverse order."))))
+
+(deftest lifted-fn-tests
+  (letfn [(breaks? [f x]
+            (is (thrown? #?(:clj Exception :cljs js/Error)
+                         ((t/gradient f) x))))]
+
+    (testing "function inputs throw, as they can't be perturbed"
+      (breaks? g/square g/sin))
+
+    (testing "function inputs throw, as they can't be perturbed"
+      (breaks? (fn [{:keys [x]}] (g/square x))
+               {:x 1}))
+
+    (checking "integer-discontinuous derivatives work" 100 [x sg/real]
+              (if (v/integral? x)
+                (do (breaks? g/floor x)
+                    (breaks? g/ceiling x)
+                    (breaks? g/integer-part x)
+                    (breaks? g/fractional-part x))
+
+                (do (is (zero? ((t/gradient g/floor) x)))
+                    (is (zero? ((t/gradient g/ceiling) x)))
+                    (is (zero? ((t/gradient g/integer-part) x)))
+                    (is (g/one? ((t/gradient g/fractional-part) x)))))))
+
+  (testing "lift-n"
+    (let [*   (t/lift-n g/* (fn [_] 1) (fn [_ y] y) (fn [x _] x))
+          Df7 (t/gradient
+               (fn x**7 [x] (* x x x x x x x)))
+          Df1 (t/gradient *)
+          Df0 (t/gradient (fn [_] (*)))]
+      (is (v/= '(* 7 (expt x 6))
+               (g/simplify (Df7 'x)))
+          "functions created with lift-n can take many args (they reduce via the
+          binary case!)")
+
+      (is (= 1 (Df1 'x))
+          "single-arity acts as id.")
+
+      (is (zero? (Df0 'x))
+          "zero-arity acts as constant")))
+
+  (testing "exercise some of the lifted fns by comparing them to numeric
+  derivatives."
+    (let [f (fn [x]
+              (g/+ (g/* (g/cos x) (g/sin x))
+                   (g/+ (g/sin x) (g/expt x 2))
+                   (g/+ (g/sin x) x)
+                   (g/log (g/abs x))))
+          Df         (t/gradient f)
+          Df-numeric (D-numeric f)]
```

Comment:
is the idea here that you don't want to pass when forward and backward are both wrong?